Sources/SourceKitLSP/Swift/ReferenceDocument.swift

```diff
@@ -0,0 +1,150 @@
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import LanguageServerProtocol
+
+/// A Reference Document is a document whose url scheme is `sourcekit-lsp:` and whose content can only be retrieved
+/// using `GetReferenceDocumentRequest`.
+///
+/// The `url` will be of the form: `sourcekit-lsp://<documentType>/<displayName>?<data>`
+/// Here,
+///  - The `documentType` denotes the kind of the content present in the reference document
+///  - The `data` denotes the parameter-value pairs such as "p1=v1&p2=v2&..." needed to generate
+/// the content of the reference document.
+///  - The `displayName` is the displayed file name of the reference document. It doesn't involve in generating
+/// the content of the reference document.
+package protocol ReferenceDocument {
+  /// A type to represent the parameter-value pairs of `data` as an object
+  associatedtype Data: ReferenceDocumentData
+
+  var url: URL { get throws }
+
+  static var documentType: ReferenceDocumentType { get }
+
+  var displayName: String { get }
+  var data: Data { get set }
+
+  init(from url: URL) throws
+  init(from urlString: String) throws
+
+  init(data: Data)
+
+  /// Creates the `DocumentURI` of the source file, using which the reference document is generated, and returns it.
+  /// - Returns: The `DocumentURI` of the source file
+  func sourceDocument() -> DocumentURI
+
+  /// This generates the content of the reference document and returns it.
+  /// - Parameter languageService: The `SwiftLanguageService` using which it generates
+  /// the content of the reference document.
+  /// - Returns: The content of the reference document
+  func documentContent(using languageService: SwiftLanguageService) async throws -> String
+}
+
+/// All custom `ReferenceDocumentType`s must be registered here along with their `ReferenceDocument.Type`.
+package var supportedReferenceDocumentTypes: [ReferenceDocumentType: any ReferenceDocument.Type] = [
+  .macroExpansion: MacroExpansionReferenceDocument.self
+]
+
+/// Represents the document type of the `ReferenceDocument`
+package enum ReferenceDocumentType: String {
+  case macroExpansion = "macro-expansion"
+}
+
+/// Represents the parameter-value pairs of the reference document's uri
+package protocol ReferenceDocumentData {
+  var queryItems: [URLQueryItem]? { get }
+
+  init(from queryItems: [URLQueryItem]?) throws
+}
+
+package extension ReferenceDocument {
+  var url: URL {
+    get throws {
+      let baseURL = "sourcekit-lsp://\(Self.documentType.rawValue)"
+      guard var components = URLComponents(string: baseURL) else {
+        throw ReferenceDocumentError(
+          description: "Unable to create URL for ReferenceDocument: \(baseURL)"
+        )
+      }
```

Comment:
Wouldn’t it be cleaner if you just create an empty `URLComponents` and then set scheme and host on it?

```suggestion
var components = URLComponents()
components.scheme = "sourcekit-lsp"
components.host = "swift-macro-expansion"
```